<!DOCTYPE html>
<html lang="en"><head><meta charset="UTF-8"/><meta name="viewport" content="width=device-width, initial-scale=1.0"/><title>WDL-GWAS on UKBiobank RAP · WDL-GWAS</title><meta name="title" content="WDL-GWAS on UKBiobank RAP · WDL-GWAS"/><meta property="og:title" content="WDL-GWAS on UKBiobank RAP · WDL-GWAS"/><meta property="twitter:title" content="WDL-GWAS on UKBiobank RAP · WDL-GWAS"/><meta name="description" content="Documentation for WDL-GWAS."/><meta property="og:description" content="Documentation for WDL-GWAS."/><meta property="twitter:description" content="Documentation for WDL-GWAS."/><meta property="og:url" content="https://olivierlabayle.github.io/WDL-GWAS/stable/tutorials/ukb_run/"/><meta property="twitter:url" content="https://olivierlabayle.github.io/WDL-GWAS/stable/tutorials/ukb_run/"/><link rel="canonical" href="https://olivierlabayle.github.io/WDL-GWAS/stable/tutorials/ukb_run/"/><script data-outdated-warner src="../../assets/warner.js"></script><link href="https://cdnjs.cloudflare.com/ajax/libs/lato-font/3.0.0/css/lato-font.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/juliamono/0.050/juliamono.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/6.4.2/css/fontawesome.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/6.4.2/css/solid.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/6.4.2/css/brands.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/KaTeX/0.16.8/katex.min.css" rel="stylesheet" type="text/css"/><script>documenterBaseURL="../.."</script><script src="https://cdnjs.cloudflare.com/ajax/libs/require.js/2.3.6/require.min.js" data-main="../../assets/documenter.js"></script><script src="../../search_index.js"></script><script src="../../siteinfo.js"></script><script src="../../../versions.js"></script><link class="docs-theme-link" rel="stylesheet" type="text/css" href="../../assets/themes/catppuccin-mocha.css" data-theme-name="catppuccin-mocha"/><link class="docs-theme-link" rel="stylesheet" type="text/css" href="../../assets/themes/catppuccin-macchiato.css" data-theme-name="catppuccin-macchiato"/><link class="docs-theme-link" rel="stylesheet" type="text/css" href="../../assets/themes/catppuccin-frappe.css" data-theme-name="catppuccin-frappe"/><link class="docs-theme-link" rel="stylesheet" type="text/css" href="../../assets/themes/catppuccin-latte.css" data-theme-name="catppuccin-latte"/><link class="docs-theme-link" rel="stylesheet" type="text/css" href="../../assets/themes/documenter-dark.css" data-theme-name="documenter-dark" data-theme-primary-dark/><link class="docs-theme-link" rel="stylesheet" type="text/css" href="../../assets/themes/documenter-light.css" data-theme-name="documenter-light" data-theme-primary/><script src="../../assets/themeswap.js"></script></head><body><div id="documenter"><nav class="docs-sidebar"><a class="docs-logo" href="../../"><img src="../../assets/logo.png" alt="WDL-GWAS logo"/></a><div class="docs-package-name"><span class="docs-autofit"><a href="../../">WDL-GWAS</a></span></div><button class="docs-search-query input is-rounded is-small is-clickable my-2 mx-auto py-1 px-2" id="documenter-search-query">Search docs (Ctrl + /)</button><ul class="docs-menu"><li><a class="tocitem" href="../../">Home</a></li><li><span class="tocitem">Tutorials</span><ul><li><a class="tocitem" href="../local_run/">WDL-GWAS Locally</a></li><li class="is-active"><a class="tocitem" href>WDL-GWAS on UKBiobank RAP</a><ul class="internal"><li><a class="tocitem" href="#Creating-the-Covariates-File"><span>Creating the Covariates File</span></a></li><li><a class="tocitem" href="#Extracting-Imputed-and-Typed-Genotypes"><span>Extracting Imputed and Typed Genotypes</span></a></li><li><a class="tocitem" href="#Running-WDL-GWAS"><span>Running WDL-GWAS</span></a></li><li><a class="tocitem" href="#Outputs"><span>Outputs</span></a></li></ul></li></ul></li><li><a class="tocitem" href="../../inputs/">Workflow&#39;s Inputs</a></li><li><a class="tocitem" href="../../outputs/">Workflow&#39;s Outputs</a></li><li><a class="tocitem" href="../../troubleshooting/">Troobleshooting</a></li><li><a class="tocitem" href="../../contributing/">Contributing and Reporting</a></li></ul><div class="docs-version-selector field has-addons"><div class="control"><span class="docs-label button is-static is-size-7">Version</span></div><div class="docs-selector control is-expanded"><div class="select is-fullwidth is-size-7"><select id="documenter-version-selector"></select></div></div></div></nav><div class="docs-main"><header class="docs-navbar"><a class="docs-sidebar-button docs-navbar-link fa-solid fa-bars is-hidden-desktop" id="documenter-sidebar-button" href="#"></a><nav class="breadcrumb"><ul class="is-hidden-mobile"><li><a class="is-disabled">Tutorials</a></li><li class="is-active"><a href>WDL-GWAS on UKBiobank RAP</a></li></ul><ul class="is-hidden-tablet"><li class="is-active"><a href>WDL-GWAS on UKBiobank RAP</a></li></ul></nav><div class="docs-right"><a class="docs-navbar-link" href="https://github.com/olivierlabayle/WDL-GWAS" title="View the repository on GitHub"><span class="docs-icon fa-brands"></span><span class="docs-label is-hidden-touch">GitHub</span></a><a class="docs-navbar-link" href="https://github.com/olivierlabayle/WDL-GWAS/blob/main/docs/src/tutorials/ukb_run.md" title="Edit source on GitHub"><span class="docs-icon fa-solid"></span></a><a class="docs-settings-button docs-navbar-link fa-solid fa-gear" id="documenter-settings-button" href="#" title="Settings"></a><a class="docs-article-toggle-button fa-solid fa-chevron-up" id="documenter-article-toggle-button" href="javascript:;" title="Collapse all docstrings"></a></div></header><article class="content" id="documenter-page"><h1 id="Running-WDL-GWAS-on-the-UK-Biobank-RAP"><a class="docs-heading-anchor" href="#Running-WDL-GWAS-on-the-UK-Biobank-RAP">Running WDL-GWAS on the UK Biobank RAP</a><a id="Running-WDL-GWAS-on-the-UK-Biobank-RAP-1"></a><a class="docs-heading-anchor-permalink" href="#Running-WDL-GWAS-on-the-UK-Biobank-RAP" title="Permalink"></a></h1><p>In this example, we provide a tutorial to reproduce the two GWAS studies performed in the WDL-GWAS paper: BMI and colorectal cancer. This involves a bit more than running the workflow since we need to prepare the data for it. Also, since you will be interacting with the RAP and real world data, this involves costs and some time investment (however most of the time is just waiting for jobs to finish so you can do something else).</p><table><tr><th style="text-align: left"><strong>Step</strong></th><th style="text-align: center"><strong>Total Time</strong></th><th style="text-align: center"><strong>Active Time</strong></th><th style="text-align: center"><strong>Expected Price</strong></th></tr><tr><td style="text-align: left">Covariates Preparation</td><td style="text-align: center">30 minutes</td><td style="text-align: center">20 minutes</td><td style="text-align: center">£0</td></tr><tr><td style="text-align: left">Genotypes Preparation</td><td style="text-align: center">4 hours</td><td style="text-align: center">10 minutes</td><td style="text-align: center">£20</td></tr><tr><td style="text-align: left">GWAS</td><td style="text-align: center">1.5 days</td><td style="text-align: center">10 minutes</td><td style="text-align: center">£130</td></tr><tr><td style="text-align: left">Total</td><td style="text-align: center">2 days</td><td style="text-align: center">40 minutes</td><td style="text-align: center">£150</td></tr></table><div class="admonition is-warning" id="Warning-8d7afa7f21fa6450"><header class="admonition-header">Warning<a class="admonition-anchor" href="#Warning-8d7afa7f21fa6450" title="Permalink"></a></header><div class="admonition-body"><p>Body mass index has many significant loci across the genome, making this run quite expensive. This is because each of them will be fine-mapped. For a cheaper run, restrict the following analysis to colorectal cancer or any trait of interest to you which should cost about £40.</p></div></div><h2 id="Creating-the-Covariates-File"><a class="docs-heading-anchor" href="#Creating-the-Covariates-File">Creating the Covariates File</a><a id="Creating-the-Covariates-File-1"></a><a class="docs-heading-anchor-permalink" href="#Creating-the-Covariates-File" title="Permalink"></a></h2><p>First we need to extract the relevant fields from the UK Biobank&#39;s main dataset, we can do this using <code>dx extract_dataset</code>:</p><pre><code class="language-bash hljs">dx extract_dataset DATASET_RECORD_ID --fields-file docs/src/assets/paper_fields.txt --output raw_covariates.csv</code></pre><div class="admonition is-warning" id="Warning-c53c7e77497ffd49"><header class="admonition-header">Warning<a class="admonition-anchor" href="#Warning-c53c7e77497ffd49" title="Permalink"></a></header><div class="admonition-body"><p>Sometimes the above fails randomly with <code>&#39;Invalid JSON received from server&#39;, 200</code>. Just try again.</p></div></div><p>However, the extracted dataset cannot be directly used with the WDL-GWAS workflow, we first need to define the covariates and phenotypes from the extracted fields. The following Julia code snippet creates such a covariate file. If you don&#39;t want to download Julia and reproduce the code, here is what it does:</p><ul><li>It makes sure the <code>FID</code> and <code>IID</code> are defined and correspond to the <code>eid</code> field.</li><li>It defines the current participants&#39; age.</li><li>It extracts the participants self reported ethnic background</li><li>It defines colorectal cancer has having any of the C18, C19 or C20 ICD10 code.</li><li>It extracts the BMI column</li></ul><p>If you want to reproduce this example, you will need to download Julia for which I recommend using <a href="https://github.com/JuliaLang/juliaup">juliaup</a>. Then, you can run the below code interactively using the REPL by typing <code>julia</code>.</p><pre><code class="language-julia hljs">using Pkg; Pkg.activate(&quot;docs&quot;); Pkg.instantiate()
using CSV
using DataFrames
using Dates

function parse_ICD10_col!(parsed_col, col, codes)
    for index in eachindex(parsed_col)
        val = col[index]
        val === missing &amp;&amp; continue
        parsed_col[index] =  parsed_col[index] | any(code -&gt; startswith(val, code), codes)
    end
    return parsed_col
end

function parse_cancer(cols...; codes)
    cancer_col = zeros(Int, length(first(cols)))
    for col in cols
        parse_ICD10_col!(cancer_col, col, codes)
    end
    return cancer_col
end

superethnicity(x::Missing) = x

&quot;&quot;&quot;
This will group `Do not know` and `Prefer not to answer` together
&quot;&quot;&quot;
superethnicity(x) = first(string(x))

cancer_registry_cols = [
    &quot;participant.p40006_i0&quot;,
    &quot;participant.p40006_i1&quot;,
    &quot;participant.p40006_i2&quot;,
    &quot;participant.p40006_i3&quot;,
    &quot;participant.p40006_i4&quot;,
    &quot;participant.p40006_i5&quot;,
    &quot;participant.p40006_i6&quot;,
    &quot;participant.p40006_i7&quot;,
    &quot;participant.p40006_i8&quot;,
    &quot;participant.p40006_i9&quot;,
    &quot;participant.p40006_i10&quot;,
    &quot;participant.p40006_i11&quot;,
    &quot;participant.p40006_i12&quot;,
    &quot;participant.p40006_i13&quot;,
    &quot;participant.p40006_i14&quot;,
    &quot;participant.p40006_i15&quot;,
    &quot;participant.p40006_i16&quot;,
    &quot;participant.p40006_i17&quot;,
    &quot;participant.p40006_i18&quot;,
    &quot;participant.p40006_i19&quot;,
    &quot;participant.p40006_i20&quot;,
    &quot;participant.p40006_i21&quot;
]

raw_covariates = CSV.read(&quot;raw_covariates.csv&quot;, DataFrame)

covariates = select(raw_covariates,
    &quot;participant.eid&quot; =&gt; &quot;FID&quot;,
    &quot;participant.eid&quot; =&gt; &quot;IID&quot;,
    &quot;participant.p22001&quot; =&gt; &quot;SEX&quot;,
    &quot;participant.p34&quot; =&gt; (x -&gt; Int(Dates.year(now())) .- x) =&gt; &quot;AGE&quot;,
    &quot;participant.p21000_i0&quot; =&gt; (x -&gt; superethnicity.(x)) =&gt; &quot;SUPER_ETHNICITY&quot;,
    &quot;participant.p23104_i0&quot; =&gt; &quot;BMI&quot;,
    cancer_registry_cols =&gt; ((cols...) -&gt; parse_cancer(cols...; codes=[&quot;C18&quot;, &quot;C19&quot;, &quot;C20&quot;])) =&gt; &quot;COLORECTAL_CANCER&quot;
)

CSV.write(&quot;covariates.csv&quot;, covariates)</code></pre><p>Finally we can upload the dataset to the UK Biobank RAP using <code>dx upload</code>:</p><pre><code class="language-bash hljs">dx upload --path /wdl_gwas_covariates.csv covariates.csv</code></pre><h2 id="Extracting-Imputed-and-Typed-Genotypes"><a class="docs-heading-anchor" href="#Extracting-Imputed-and-Typed-Genotypes">Extracting Imputed and Typed Genotypes</a><a id="Extracting-Imputed-and-Typed-Genotypes-1"></a><a class="docs-heading-anchor-permalink" href="#Extracting-Imputed-and-Typed-Genotypes" title="Permalink"></a></h2><p>WDL-GWAS requires both imputed <strong>and</strong> typed genotypes. To get those we will (i) use the TOPMed imputed genotypes which should be provided in your <code>/Bulk/Imputation/</code> folder and (ii) add some quality control filters. And of course, to make this reproducible, we will use another WDL workflow.</p><p>The workflow can be found in <code>workflows/extract_genotypes.wdl</code>, feel free to adapt it to your own needs.</p><p>Then you can run it using the following DNA Nexus commands. First we compile it using dxCompiler:</p><pre><code class="nohighlight hljs">java -jar $DX_COMPILER_PATH compile workflows/extract_genotypes.wdl \
-f -project $RAP_PROJECT_ID \
-reorg \
-folder /workflows/extract_genotypes \
-inputs docs/src/assets/extract-ukb-genotypes.inputs.json</code></pre><p>where:</p><ul><li><code>DX_COMPILER_PATH</code> is set as per the installation&#39;s instruction.</li><li><code>RAP_PROJECT_ID</code> is your UK Biobank RAP project ID.</li><li><code>docs/src/assets/extract-ukb-genotypes.inputs.json</code> is the file providing the workflow&#39;s inputs. In particular it points to the various BGEN files from TOPMed and VCF sites files. </li></ul><p>The compiler might output some warnings like <code>missing input for non-optional parameter</code> but you can ignore these.</p><p>The command will do a few things:</p><ol><li>It will compile and upload the workflow to the UK Biobank RAP </li><li>It will create a corresponding <code>inputs.dx.json</code> file locally which can be used to run the workflow </li></ol><div class="admonition is-info" id="Note-bf31089fd55cd175"><header class="admonition-header">Note<a class="admonition-anchor" href="#Note-bf31089fd55cd175" title="Permalink"></a></header><div class="admonition-body"><p>File paths need to point to their location in your UK Biobank RAP project. There are two main ways you can fill those in - via object identifiers: <code>dx://RAP_PROJECT_ID:FILE_ID</code> or via a file path: <code>dx://RAP_PROJECT_ID:/path/to/file</code>.</p></div></div><p>Then we can run the workflow with:</p><pre><code class="nohighlight hljs">dx run -y \
-f docs/src/assets/extract-ukb-genotypes.inputs.dx.json \
--priority high \
--destination /ukb_extracted_genotypes/ \
/workflows/extract_genotypes/extract_ukb_genotypes</code></pre><p>This workflow runs in approximately 5 hours and will cost about £20. The resulting files will be available in the <code>/ukb_extracted_genotypes</code> folder.</p><h2 id="Running-WDL-GWAS"><a class="docs-heading-anchor" href="#Running-WDL-GWAS">Running WDL-GWAS</a><a id="Running-WDL-GWAS-1"></a><a class="docs-heading-anchor-permalink" href="#Running-WDL-GWAS" title="Permalink"></a></h2><p>Similarly to the previous step, we need to provide WDL-GWAS&#39; inputs via a JSON file. In this example, the file paths will point to the freshly extracted imputed and typed genotypes. We also need to provide some phenotypes of interest, here <code>BMI</code> and <code>COLORECTAL_CANCER</code> and for illustration, we will restrict the analysis to a self-reported White ethnic background.</p><p>The full JSON input file is located in <code>docs/src/assets/paper.inputs.json</code> for which an extract is presented here:</p><pre><code class="language-json hljs">{
    &quot;gwas.covariates_file&quot;: &quot;dx://project-J0pkqyQJpYQ133JG1p2J1qzv:/wdl_gwas_covariates.csv&quot;,
    &quot;gwas.genotypes&quot;: {
        &quot;chr&quot;: &quot;all&quot;,
        &quot;bed&quot;: &quot;dx://project-J0pkqyQJpYQ133JG1p2J1qzv:/ukb_extracted_genotypes/typed.bed&quot;,
        &quot;bim&quot;: &quot;dx://project-J0pkqyQJpYQ133JG1p2J1qzv:/ukb_extracted_genotypes/typed.bim&quot;,
        &quot;fam&quot;: &quot;dx://project-J0pkqyQJpYQ133JG1p2J1qzv:/ukb_extracted_genotypes/typed.fam&quot;
    },
    &quot;gwas.imputed_genotypes&quot;: [
        {
            &quot;chr&quot;: &quot;1&quot;,
            &quot;pgen&quot;: &quot;dx://project-J0pkqyQJpYQ133JG1p2J1qzv:/ukb_extracted_genotypes/ukb21007_c1_b0_v1.imputed.pgen&quot;,
            &quot;psam&quot;: &quot;dx://project-J0pkqyQJpYQ133JG1p2J1qzv:/ukb_extracted_genotypes/ukb21007_c1_b0_v1.imputed.psam&quot;,
            &quot;pvar&quot;: &quot;dx://project-J0pkqyQJpYQ133JG1p2J1qzv:/ukb_extracted_genotypes/ukb21007_c1_b0_v1.imputed.pvar&quot;
        },
        {
            &quot;chr&quot;: &quot;2&quot;,
            &quot;pgen&quot;: &quot;dx://project-J0pkqyQJpYQ133JG1p2J1qzv:/ukb_extracted_genotypes/ukb21007_c2_b0_v1.imputed.pgen&quot;,
            &quot;psam&quot;: &quot;dx://project-J0pkqyQJpYQ133JG1p2J1qzv:/ukb_extracted_genotypes/ukb21007_c2_b0_v1.imputed.psam&quot;,
            &quot;pvar&quot;: &quot;dx://project-J0pkqyQJpYQ133JG1p2J1qzv:/ukb_extracted_genotypes/ukb21007_c2_b0_v1.imputed.pvar&quot;
        },
    ...
    ],
    &quot;gwas.covariates&quot;: [&quot;AGE&quot;, &quot;SEX&quot;, &quot;AGE_x_AGE&quot;, &quot;AGE_x_SEX&quot;],
    &quot;gwas.phenotypes&quot;: [&quot;COLORECTAL_CANCER&quot;, &quot;BMI&quot;],
    &quot;gwas.filterby&quot;: [&quot;SUPER_ETHNICITY=1&quot;],
    ...
}</code></pre><p>Then, as in the previous step, we need to compile WDL-GWAS and upload it to the RAP, this can be done with <a href="https://github.com/dnanexus/dxCompiler/blob/develop/doc/ExpertOptions.md">dxCompiler</a>:</p><pre><code class="language-bash hljs">java -jar $DX_COMPILER_PATH compile workflows/gwas.wdl \
-f -project $RAP_PROJECT_ID \
-reorg \
-folder /workflows/gwas \
-inputs docs/src/assets/paper.inputs.json</code></pre><p>Then, to run WDL-GWAS:</p><pre><code class="language-bash hljs">dx run -y \
-f docs/src/assets/paper.inputs.dx.json \
--priority high \
--destination /wdl_gwas_paper_outputs/ \
/workflows/gwas/gwas</code></pre><p>Of course you can change the destination <code>/wdl_gwas_paper_outputs</code> to your need. You should be able to visualize the run execution on your UK Biobank RAP account under the <code>MONITOR</code> tab.</p><h2 id="Outputs"><a class="docs-heading-anchor" href="#Outputs">Outputs</a><a id="Outputs-1"></a><a class="docs-heading-anchor-permalink" href="#Outputs" title="Permalink"></a></h2><p>The workflow outputs will be stored in the <code>/wdl_gwas_paper_outputs/</code> folder. </p><ul><li>In particular, the asscociated summary statistics can be found in <code>all.TRAIT.gwas.tsv</code> where <code>TRAIT</code> is either <code>BMI</code> or <code>COLORECTAL_CANCER</code>. The <code>all</code> prefix indicates that no groups were provided. Similarly, Manhattan and QQ plots are named <code>all.TRAIT.manhattan.png</code> and <code>all.TRAIT.qq.png</code> and displayed below as an illustration.</li></ul><p><img src="../../assets/outputs.png" alt="WDL GWAS Outputs"/></p><ul><li>Fine-mapping results are stored in files named <code>all.TRAIT.finemapping.tsv</code> with <code>TRAIT</code> being either <code>BMI</code> or <code>COLORECTAL_CANCER</code> as before. Furthermore, a locus plot will be created for each fine-mapped locus. These plots are named <code>all.TRAIT.LEAD_VARIANT_ID.locuszoom.png</code> where the <code>LEAD_VARIANT_ID</code> tags the locus. For instance, the locus tagged by the classic FTO variant rs1421085 is found in <code>all.BMI.16_53767042_T_C.locuszoom.png</code>. It has been assigned to a credible set with posterior inclusion probability 0.35, the highest in the set.</li></ul><p><img src="../../assets/BMI.rs1421085.locuszoom.png" alt="rs1421085"/></p></article><nav class="docs-footer"><a class="docs-footer-prevpage" href="../local_run/">« WDL-GWAS Locally</a><a class="docs-footer-nextpage" href="../../inputs/">Workflow&#39;s Inputs »</a><div class="flexbox-break"></div><p class="footer-message">Powered by <a href="https://github.com/JuliaDocs/Documenter.jl">Documenter.jl</a> and the <a href="https://julialang.org/">Julia Programming Language</a>.</p></nav></div><div class="modal" id="documenter-settings"><div class="modal-background"></div><div class="modal-card"><header class="modal-card-head"><p class="modal-card-title">Settings</p><button class="delete"></button></header><section class="modal-card-body"><p><label class="label">Theme</label><div class="select"><select id="documenter-themepicker"><option value="auto">Automatic (OS)</option><option value="documenter-light">documenter-light</option><option value="documenter-dark">documenter-dark</option><option value="catppuccin-latte">catppuccin-latte</option><option value="catppuccin-frappe">catppuccin-frappe</option><option value="catppuccin-macchiato">catppuccin-macchiato</option><option value="catppuccin-mocha">catppuccin-mocha</option></select></div></p><hr/><p>This document was generated with <a href="https://github.com/JuliaDocs/Documenter.jl">Documenter.jl</a> version 1.14.1 on <span class="colophon-date" title="Friday 24 October 2025 11:00">Friday 24 October 2025</span>. Using Julia version 1.12.1.</p></section><footer class="modal-card-foot"></footer></div></div></div></body></html>
