<!DOCTYPE html>
<html lang="en"><head><meta charset="UTF-8"/><meta name="viewport" content="width=device-width, initial-scale=1.0"/><title>WDL-GWAS on UKBiobank RAP · WDL-GWAS</title><meta name="title" content="WDL-GWAS on UKBiobank RAP · WDL-GWAS"/><meta property="og:title" content="WDL-GWAS on UKBiobank RAP · WDL-GWAS"/><meta property="twitter:title" content="WDL-GWAS on UKBiobank RAP · WDL-GWAS"/><meta name="description" content="Documentation for WDL-GWAS."/><meta property="og:description" content="Documentation for WDL-GWAS."/><meta property="twitter:description" content="Documentation for WDL-GWAS."/><meta property="og:url" content="https://olivierlabayle.github.io/WDL-GWAS/stable/tutorials/ukb_run/"/><meta property="twitter:url" content="https://olivierlabayle.github.io/WDL-GWAS/stable/tutorials/ukb_run/"/><link rel="canonical" href="https://olivierlabayle.github.io/WDL-GWAS/stable/tutorials/ukb_run/"/><script data-outdated-warner src="../../assets/warner.js"></script><link href="https://cdnjs.cloudflare.com/ajax/libs/lato-font/3.0.0/css/lato-font.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/juliamono/0.050/juliamono.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/6.4.2/css/fontawesome.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/6.4.2/css/solid.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/6.4.2/css/brands.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/KaTeX/0.16.8/katex.min.css" rel="stylesheet" type="text/css"/><script>documenterBaseURL="../.."</script><script src="https://cdnjs.cloudflare.com/ajax/libs/require.js/2.3.6/require.min.js" data-main="../../assets/documenter.js"></script><script src="../../search_index.js"></script><script src="../../siteinfo.js"></script><script src="../../../versions.js"></script><link class="docs-theme-link" rel="stylesheet" type="text/css" href="../../assets/themes/catppuccin-mocha.css" data-theme-name="catppuccin-mocha"/><link class="docs-theme-link" rel="stylesheet" type="text/css" href="../../assets/themes/catppuccin-macchiato.css" data-theme-name="catppuccin-macchiato"/><link class="docs-theme-link" rel="stylesheet" type="text/css" href="../../assets/themes/catppuccin-frappe.css" data-theme-name="catppuccin-frappe"/><link class="docs-theme-link" rel="stylesheet" type="text/css" href="../../assets/themes/catppuccin-latte.css" data-theme-name="catppuccin-latte"/><link class="docs-theme-link" rel="stylesheet" type="text/css" href="../../assets/themes/documenter-dark.css" data-theme-name="documenter-dark" data-theme-primary-dark/><link class="docs-theme-link" rel="stylesheet" type="text/css" href="../../assets/themes/documenter-light.css" data-theme-name="documenter-light" data-theme-primary/><script src="../../assets/themeswap.js"></script></head><body><div id="documenter"><nav class="docs-sidebar"><a class="docs-logo" href="../../"><img src="../../assets/logo.png" alt="WDL-GWAS logo"/></a><div class="docs-package-name"><span class="docs-autofit"><a href="../../">WDL-GWAS</a></span></div><button class="docs-search-query input is-rounded is-small is-clickable my-2 mx-auto py-1 px-2" id="documenter-search-query">Search docs (Ctrl + /)</button><ul class="docs-menu"><li><a class="tocitem" href="../../">Home</a></li><li><span class="tocitem">Tutorials</span><ul><li><a class="tocitem" href="../local_run/">WDL-GWAS Locally</a></li><li class="is-active"><a class="tocitem" href>WDL-GWAS on UKBiobank RAP</a><ul class="internal"><li><a class="tocitem" href="#Create-the-Covariates-File"><span>Create the Covariates File</span></a></li><li><a class="tocitem" href="#Extracting-Imputed-and-Typed-Genotypes"><span>Extracting Imputed and Typed Genotypes</span></a></li><li><a class="tocitem" href="#Describing-the-Inputs"><span>Describing the Inputs</span></a></li><li><a class="tocitem" href="#Running-the-Workflow"><span>Running the Workflow</span></a></li><li><a class="tocitem" href="#Outputs"><span>Outputs</span></a></li></ul></li></ul></li><li><a class="tocitem" href="../../inputs/">Workflow&#39;s Inputs</a></li><li><a class="tocitem" href="../../outputs/">Workflow&#39;s Outputs</a></li><li><a class="tocitem" href="../../troubleshooting/">Troobleshooting</a></li><li><a class="tocitem" href="../../contributing/">Contributing and Reporting</a></li></ul><div class="docs-version-selector field has-addons"><div class="control"><span class="docs-label button is-static is-size-7">Version</span></div><div class="docs-selector control is-expanded"><div class="select is-fullwidth is-size-7"><select id="documenter-version-selector"></select></div></div></div></nav><div class="docs-main"><header class="docs-navbar"><a class="docs-sidebar-button docs-navbar-link fa-solid fa-bars is-hidden-desktop" id="documenter-sidebar-button" href="#"></a><nav class="breadcrumb"><ul class="is-hidden-mobile"><li><a class="is-disabled">Tutorials</a></li><li class="is-active"><a href>WDL-GWAS on UKBiobank RAP</a></li></ul><ul class="is-hidden-tablet"><li class="is-active"><a href>WDL-GWAS on UKBiobank RAP</a></li></ul></nav><div class="docs-right"><a class="docs-navbar-link" href="https://github.com/olivierlabayle/WDL-GWAS" title="View the repository on GitHub"><span class="docs-icon fa-brands"></span><span class="docs-label is-hidden-touch">GitHub</span></a><a class="docs-navbar-link" href="https://github.com/olivierlabayle/WDL-GWAS/blob/main/docs/src/tutorials/ukb_run.md" title="Edit source on GitHub"><span class="docs-icon fa-solid"></span></a><a class="docs-settings-button docs-navbar-link fa-solid fa-gear" id="documenter-settings-button" href="#" title="Settings"></a><a class="docs-article-toggle-button fa-solid fa-chevron-up" id="documenter-article-toggle-button" href="javascript:;" title="Collapse all docstrings"></a></div></header><article class="content" id="documenter-page"><h1 id="Running-WDL-GWAS-on-the-UK-Biobank-RAP"><a class="docs-heading-anchor" href="#Running-WDL-GWAS-on-the-UK-Biobank-RAP">Running WDL-GWAS on the UK Biobank RAP</a><a id="Running-WDL-GWAS-on-the-UK-Biobank-RAP-1"></a><a class="docs-heading-anchor-permalink" href="#Running-WDL-GWAS-on-the-UK-Biobank-RAP" title="Permalink"></a></h1><div class="admonition is-info" id="Note-513b86b5faeab5ef"><header class="admonition-header">Note<a class="admonition-anchor" href="#Note-513b86b5faeab5ef" title="Permalink"></a></header><div class="admonition-body"><p>This page is under development!</p></div></div><p>In this example, we provide a tutorial to reproduce the two GWAS studies performed in the WDL-GWAS paper: BMI and colorectal cancer. This involves a bit more than running the workflow since we need to prepare the data for it.</p><h2 id="Create-the-Covariates-File"><a class="docs-heading-anchor" href="#Create-the-Covariates-File">Create the Covariates File</a><a id="Create-the-Covariates-File-1"></a><a class="docs-heading-anchor-permalink" href="#Create-the-Covariates-File" title="Permalink"></a></h2><p>First we need to extract the relevant fields from the UK Biobank&#39;s main dataset, we can do this using <code>dx extract_dataset</code>:</p><pre><code class="language-bash hljs">dx extract_dataset DATASET_RECORD_ID --fields-file docs/src/assets/paper_fields.txt --output raw_covariates.csv</code></pre><p>However, the extracted dataset cannot be directly used with the WDL-GWAS workflow, we first need to define the covariates and phenotypes from the extracted fields. The following Julia code snippet creates such a covariate file. Here is what id does:</p><ul><li>It makes sure the <code>FID</code> and <code>IID</code> are defined and correspond to the <code>eid</code> field.</li><li>It defines the current participants&#39; age.</li><li>It defines colorectal cancer has having any of the C18, C19 or C20 ICD10 code.</li><li>It extracts the BMI column</li></ul><pre><code class="language-julia hljs">using CSV
using DataFrames
using Dates

function parse_ICD10_col!(parsed_col, col, codes)
    for index in eachindex(parsed_col)
        val = col[index]
        val === missing &amp;&amp; continue
        parsed_col[index] =  parsed_col[index] | any(code -&gt; startswith(val, code), codes)
    end
    return parsed_col
end

function parse_cancer(cols...; codes)
    cancer_col = zeros(Int, length(first(cols)))
    for col in cols
        parse_ICD10_col!(cancer_col, col, codes)
    end
    return cancer_col
end

cancer_registry_cols = [
    &quot;participant.p40006_i0&quot;,
    &quot;participant.p40006_i1&quot;,
    &quot;participant.p40006_i2&quot;,
    &quot;participant.p40006_i3&quot;,
    &quot;participant.p40006_i4&quot;,
    &quot;participant.p40006_i5&quot;,
    &quot;participant.p40006_i6&quot;,
    &quot;participant.p40006_i7&quot;,
    &quot;participant.p40006_i8&quot;,
    &quot;participant.p40006_i9&quot;,
    &quot;participant.p40006_i10&quot;,
    &quot;participant.p40006_i11&quot;,
    &quot;participant.p40006_i12&quot;,
    &quot;participant.p40006_i13&quot;,
    &quot;participant.p40006_i14&quot;,
    &quot;participant.p40006_i15&quot;,
    &quot;participant.p40006_i16&quot;,
    &quot;participant.p40006_i17&quot;,
    &quot;participant.p40006_i18&quot;,
    &quot;participant.p40006_i19&quot;,
    &quot;participant.p40006_i20&quot;,
    &quot;participant.p40006_i21&quot;
]

raw_covariates = CSV.read(&quot;raw_covariates.csv&quot;, DataFrame)

covariates = select(raw_covariates,
    &quot;participant.eid&quot; =&gt; &quot;FID&quot;,
    &quot;participant.eid&quot; =&gt; &quot;IID&quot;,
    &quot;participant.p22001&quot; =&gt; &quot;SEX&quot;,
    &quot;participant.p34&quot; =&gt; (x -&gt; Int(Dates.year(now())) .- x) =&gt; &quot;AGE&quot;,
    &quot;participant.p23104_i0&quot; =&gt; &quot;BMI&quot;,
    cancer_registry_cols =&gt; ((cols...) -&gt; parse_cancer(cols...; codes=[&quot;C18&quot;, &quot;C19&quot;, &quot;C20&quot;])) =&gt; &quot;COLORECTAL_CANCER&quot;
)

CSV.write(&quot;covariates.csv&quot;, covariates)</code></pre><p>Finally we can upload the dataset to the UK Biobank RAP using <code>dx upload</code>:</p><pre><code class="language-bash hljs">dx upload --path /wdl_gwas_covariates.csv covariates.csv</code></pre><h2 id="Extracting-Imputed-and-Typed-Genotypes"><a class="docs-heading-anchor" href="#Extracting-Imputed-and-Typed-Genotypes">Extracting Imputed and Typed Genotypes</a><a id="Extracting-Imputed-and-Typed-Genotypes-1"></a><a class="docs-heading-anchor-permalink" href="#Extracting-Imputed-and-Typed-Genotypes" title="Permalink"></a></h2><p>WDL-GWAS requires both imputed <strong>and</strong> typed genotypes. To get those we will (i) use the TOPMed imputed genotypes which should be provided in your <code>/Bulk/Imputation/</code> folder and (ii) add some quality control filters. And of course, to make this reproducible, we will use another WDL workflow.</p><p>The workflow can be found in <code>workflows/extract_genotypes.wdl</code>, feel free to adapt it to your own needs.</p><p>Then you can run it using the following DNA Nexus commands. First we compile it using dxCompiler:</p><pre><code class="nohighlight hljs">java -jar $DX_COMPILER_PATH compile workflows/extract_genotypes.wdl \
-f -project $RAP_PROJECT_ID \
-reorg \
-folder /workflows/extract_genotypes \
-inputs docs/src/assets/extract-ukb-genotypes.inputs.json</code></pre><p>The <code>docs/src/assets/extract-ukb-genotypes.inputs.json</code> points to the various BGEN files from TOPMed and VCF sites files.</p><p>Then we can run it with:</p><pre><code class="nohighlight hljs">dx run -y \
-f docs/src/assets/extract-ukb-genotypes.inputs.dx.json \
--priority high \
--destination /ukb_extracted_genotypes/ \
/workflows/extract_genotypes/extract_ukb_genotypes</code></pre><h2 id="Describing-the-Inputs"><a class="docs-heading-anchor" href="#Describing-the-Inputs">Describing the Inputs</a><a id="Describing-the-Inputs-1"></a><a class="docs-heading-anchor-permalink" href="#Describing-the-Inputs" title="Permalink"></a></h2><p>As per the <a href="../local_run/#Running-WDL-GWAS-Locally">Running WDL-GWAS Locally</a> example, the workflow&#39;s inputs are provided via a JSON file. However this time, the file paths need to point to teir location in your UK Biobank RAP project. There are two main ways you can fill those in:</p><ul><li>Via object identifiers: <code>dx://RAP_PROJECT_ID:FILE_ID</code></li><li>Via a file path: <code>dx://RAP_PROJECT_ID:/path/to/file</code></li></ul><p>An ecample inputs JSON file could look like:</p><pre><code class="language-json hljs">{
    &quot;gwas.meta_exclude&quot;: [&quot;AMR&quot;],
    &quot;gwas.covariates_file&quot;: &quot;dx://RAP_PROJECT_ID:/path/to/covariates/file&quot;,
    &quot;gwas.groupby&quot;: [&quot;SUPERPOPULATION&quot;],
    &quot;gwas.covariates&quot;: [&quot;AGE&quot;, &quot;SEX&quot;, &quot;AGE_x_AGE&quot;],
    &quot;gwas.phenotypes&quot;: [&quot;SEVERE_PNEUMONIA&quot;, &quot;SEVERE_COVID_19&quot;],
    &quot;gwas.genotypes&quot;: {
        &quot;chr&quot;: &quot;all&quot;,
        &quot;bed&quot;: &quot;dx://RAP_PROJECT_ID:/path/to/bed/file&quot;,
        &quot;bim&quot;: &quot;dx://RAP_PROJECT_ID:/path/to/bim/file&quot;,
        &quot;fam&quot;: &quot;dx://RAP_PROJECT_ID:/path/to/fam/file&quot;
    },
    &quot;gwas.imputed_genotypes&quot;: [
        {
            &quot;chr&quot;: &quot;1&quot;,
            &quot;pgen&quot;: &quot;dx://RAP_PROJECT_ID:PGEN_CHR_1_FILE_ID&quot;,
            &quot;psam&quot;: &quot;dx://RAP_PROJECT_ID:PSAM_CHR_1_FILE_ID&quot;,
            &quot;pvar&quot;: &quot;dx://RAP_PROJECT_ID:PVAR_CHR_1_FILE_ID&quot;
        },
        {
            &quot;chr&quot;: &quot;2&quot;,
            &quot;pgen&quot;: &quot;dx://RAP_PROJECT_ID:PGEN_CHR_2_FILE_ID&quot;,
            &quot;psam&quot;: &quot;dx://RAP_PROJECT_ID:PSAM_CHR_2_FILE_ID&quot;,
            &quot;pvar&quot;: &quot;dx://RAP_PROJECT_ID:PVAR_CHR_2_FILE_ID&quot;
        },
        {
            &quot;chr&quot;: &quot;3&quot;,
            &quot;pgen&quot;: &quot;dx://RAP_PROJECT_ID:PGEN_CHR_3_FILE_ID&quot;,
            &quot;psam&quot;: &quot;dx://RAP_PROJECT_ID:PSAM_CHR_3_FILE_ID&quot;,
            &quot;pvar&quot;: &quot;dx://RAP_PROJECT_ID:PVAR_CHR_3_FILE_ID&quot;
        },
    ],
}</code></pre><h2 id="Running-the-Workflow"><a class="docs-heading-anchor" href="#Running-the-Workflow">Running the Workflow</a><a id="Running-the-Workflow-1"></a><a class="docs-heading-anchor-permalink" href="#Running-the-Workflow" title="Permalink"></a></h2><p>First you need to compile the workflow and upload it to the RAP, this can be done with <a href="https://github.com/dnanexus/dxCompiler/blob/develop/doc/ExpertOptions.md">dxCompiler</a>:</p><pre><code class="language-bash hljs">java -jar $DX_COMPILER_PATH compile workflows/gwas.wdl \
-f -project $RAP_PROJECT_ID \
-reorg \
-folder /workflows/gwas \
-inputs inputs.json</code></pre><p>where:</p><ul><li><code>DX_COMPILER_PATH</code> is set as per the installation&#39;s instruction.</li><li><code>RAP_PROJECT_ID</code> is your UK Biobank RAP project ID.</li><li><code>inputs.json</code> is your workflow&#39;s input file. </li></ul><p>The compiler might output some warnings like <code>missing input for non-optional parameter</code> but you can ignore these.</p><p>The command will do a few things:</p><ol><li>It will compile and upload the workflow to the UK Biobank RAP </li><li>It will create a corresponding <code>inputs.dx.json</code> file locally which can be used to run the workflow </li></ol><p>Then, to run the workflow, run:</p><pre><code class="language-bash hljs">dx run -y \
-f inputs.dx.json \
--priority high \
--destination /workflow_outputs/ \
/workflows/gwas/gwas</code></pre><p>Of course you can change the destination <code>/workflow_outputs</code> to your need. You should be able to visualize the run execution on your UK Biobank RAP account under the <code>MONITOR</code> tab.</p><h2 id="Outputs"><a class="docs-heading-anchor" href="#Outputs">Outputs</a><a id="Outputs-1"></a><a class="docs-heading-anchor-permalink" href="#Outputs" title="Permalink"></a></h2><p>Outputs will be stored in the <code>/workflow_outputs//</code> folder. Since the RAP preserves intermediate files, this will be quite dense.</p><ul><li>plots: searching for <code>png</code></li><li>group/phenotype results: searching for <code>regenie.results.${group_name}.${phenotype}.tsv</code></li></ul></article><nav class="docs-footer"><a class="docs-footer-prevpage" href="../local_run/">« WDL-GWAS Locally</a><a class="docs-footer-nextpage" href="../../inputs/">Workflow&#39;s Inputs »</a><div class="flexbox-break"></div><p class="footer-message">Powered by <a href="https://github.com/JuliaDocs/Documenter.jl">Documenter.jl</a> and the <a href="https://julialang.org/">Julia Programming Language</a>.</p></nav></div><div class="modal" id="documenter-settings"><div class="modal-background"></div><div class="modal-card"><header class="modal-card-head"><p class="modal-card-title">Settings</p><button class="delete"></button></header><section class="modal-card-body"><p><label class="label">Theme</label><div class="select"><select id="documenter-themepicker"><option value="auto">Automatic (OS)</option><option value="documenter-light">documenter-light</option><option value="documenter-dark">documenter-dark</option><option value="catppuccin-latte">catppuccin-latte</option><option value="catppuccin-frappe">catppuccin-frappe</option><option value="catppuccin-macchiato">catppuccin-macchiato</option><option value="catppuccin-mocha">catppuccin-mocha</option></select></div></p><hr/><p>This document was generated with <a href="https://github.com/JuliaDocs/Documenter.jl">Documenter.jl</a> version 1.14.1 on <span class="colophon-date" title="Wednesday 8 October 2025 15:58">Wednesday 8 October 2025</span>. Using Julia version 1.12.0.</p></section><footer class="modal-card-foot"></footer></div></div></div></body></html>
