<!DOCTYPE html>
<html lang="en"><head><meta charset="UTF-8"/><meta name="viewport" content="width=device-width, initial-scale=1.0"/><title>Workflow Parameters · WDL-GWAS</title><meta name="title" content="Workflow Parameters · WDL-GWAS"/><meta property="og:title" content="Workflow Parameters · WDL-GWAS"/><meta property="twitter:title" content="Workflow Parameters · WDL-GWAS"/><meta name="description" content="Documentation for WDL-GWAS."/><meta property="og:description" content="Documentation for WDL-GWAS."/><meta property="twitter:description" content="Documentation for WDL-GWAS."/><meta property="og:url" content="https://olivierlabayle.github.io/WDL-GWAS/stable/parameters/"/><meta property="twitter:url" content="https://olivierlabayle.github.io/WDL-GWAS/stable/parameters/"/><link rel="canonical" href="https://olivierlabayle.github.io/WDL-GWAS/stable/parameters/"/><script data-outdated-warner src="../assets/warner.js"></script><link href="https://cdnjs.cloudflare.com/ajax/libs/lato-font/3.0.0/css/lato-font.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/juliamono/0.050/juliamono.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/6.4.2/css/fontawesome.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/6.4.2/css/solid.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/6.4.2/css/brands.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/KaTeX/0.16.8/katex.min.css" rel="stylesheet" type="text/css"/><script>documenterBaseURL=".."</script><script src="https://cdnjs.cloudflare.com/ajax/libs/require.js/2.3.6/require.min.js" data-main="../assets/documenter.js"></script><script src="../search_index.js"></script><script src="../siteinfo.js"></script><script src="../../versions.js"></script><link class="docs-theme-link" rel="stylesheet" type="text/css" href="../assets/themes/catppuccin-mocha.css" data-theme-name="catppuccin-mocha"/><link class="docs-theme-link" rel="stylesheet" type="text/css" href="../assets/themes/catppuccin-macchiato.css" data-theme-name="catppuccin-macchiato"/><link class="docs-theme-link" rel="stylesheet" type="text/css" href="../assets/themes/catppuccin-frappe.css" data-theme-name="catppuccin-frappe"/><link class="docs-theme-link" rel="stylesheet" type="text/css" href="../assets/themes/catppuccin-latte.css" data-theme-name="catppuccin-latte"/><link class="docs-theme-link" rel="stylesheet" type="text/css" href="../assets/themes/documenter-dark.css" data-theme-name="documenter-dark" data-theme-primary-dark/><link class="docs-theme-link" rel="stylesheet" type="text/css" href="../assets/themes/documenter-light.css" data-theme-name="documenter-light" data-theme-primary/><script src="../assets/themeswap.js"></script></head><body><div id="documenter"><nav class="docs-sidebar"><a class="docs-logo" href="../"><img src="../assets/logo.png" alt="WDL-GWAS logo"/></a><div class="docs-package-name"><span class="docs-autofit"><a href="../">WDL-GWAS</a></span></div><button class="docs-search-query input is-rounded is-small is-clickable my-2 mx-auto py-1 px-2" id="documenter-search-query">Search docs (Ctrl + /)</button><ul class="docs-menu"><li><a class="tocitem" href="../">Home</a></li><li><span class="tocitem">Tutorials</span><ul><li><a class="tocitem" href="../tutorials/local_run/">WDL-GWAS Locally</a></li><li><a class="tocitem" href="../tutorials/ukb_run/">WDL-GWAS on UKBiobank RAP</a></li></ul></li><li class="is-active"><a class="tocitem" href>Workflow Parameters</a></li><li><a class="tocitem" href="../troubleshooting/">Troobleshooting</a></li><li><a class="tocitem" href="../contributing/">Contributing and Reporting</a></li></ul><div class="docs-version-selector field has-addons"><div class="control"><span class="docs-label button is-static is-size-7">Version</span></div><div class="docs-selector control is-expanded"><div class="select is-fullwidth is-size-7"><select id="documenter-version-selector"></select></div></div></div></nav><div class="docs-main"><header class="docs-navbar"><a class="docs-sidebar-button docs-navbar-link fa-solid fa-bars is-hidden-desktop" id="documenter-sidebar-button" href="#"></a><nav class="breadcrumb"><ul class="is-hidden-mobile"><li class="is-active"><a href>Workflow Parameters</a></li></ul><ul class="is-hidden-tablet"><li class="is-active"><a href>Workflow Parameters</a></li></ul></nav><div class="docs-right"><a class="docs-navbar-link" href="https://github.com/olivierlabayle/WDL-GWAS" title="View the repository on GitHub"><span class="docs-icon fa-brands"></span><span class="docs-label is-hidden-touch">GitHub</span></a><a class="docs-navbar-link" href="https://github.com/olivierlabayle/WDL-GWAS/blob/main/docs/src/parameters.md" title="Edit source on GitHub"><span class="docs-icon fa-solid"></span></a><a class="docs-settings-button docs-navbar-link fa-solid fa-gear" id="documenter-settings-button" href="#" title="Settings"></a><a class="docs-article-toggle-button fa-solid fa-chevron-up" id="documenter-article-toggle-button" href="javascript:;" title="Collapse all docstrings"></a></div></header><article class="content" id="documenter-page"><h1 id="Workflow-Parameters"><a class="docs-heading-anchor" href="#Workflow-Parameters">Workflow Parameters</a><a id="Workflow-Parameters-1"></a><a class="docs-heading-anchor-permalink" href="#Workflow-Parameters" title="Permalink"></a></h1><ul><li><code>docker_image</code> (default: olivierlabayle/wdl-gwas:main): The docker image used to run the workflow.</li><li><code>covariates_file</code>: A CSV file containing the set of covariates and phenotypes. Missing values should be represented by empty fields.</li><li><code>genotypes</code>: PLINK BED files.</li><li><code>imputed_genotypes</code>: PLINK PGEN files, split by chromosome.</li><li><code>groupby</code> (default: []): A set of variables used to stratify individuals for which a GWAS will be run independently. If empty, the full dataset is used.</li><li><code>covariates</code> (default: [&quot;AGE&quot;, &quot;SEX&quot;, &quot;AGE<em>x</em>AGE&quot;, &quot;AGE<em>x</em>SEX&quot;]): A set of covariates used to adjust for confounding or increase power in the association testing step. Product of variables can be defined using the <code>_x_</code> syntax, for example: [&quot;AGE&quot;, &quot;SEX&quot;, &quot;AGE<em>x</em>SEX&quot;, &quot;AGE<em>x</em>AGE&quot;].</li><li><code>phenotypes</code>: A set of phenotypes for which a GWAS will be run independently.</li><li><code>min_cases_controls</code> (default: 10): Minimum number of phenotype cases/controls within a group to proceed to GWAS.</li><li><code>high_ld_regions</code>: File containing high LD regions to be excluded when performing LD pruning for PCA. The file is stored in <code>assets/exclude_b38.txt</code> and needs to be uploaded to the RAP.</li><li><code>ip_values</code> (default: &quot;1000 50 0.05&quot;): Values used to create independent genotypes for PCA (see <a href="https://www.cog-genomics.org/plink/2.0/ld">here</a>).</li><li><code>npcs</code> (default 10): Number of principal components to use to account for population structure.</li><li><code>approx_pca</code> (default: true): Whether to use an approximation to the PCA algorithm (see <a href="https://www.cog-genomics.org/plink/2.0/strat">here</a>).</li><li><code>maf</code> (default: 0.01): Minor allele frequency threshold. This is used to filter variants for PCA and for plotting GWAS results. The association testing step is still performed across all variants.</li><li><code>mac</code> (default: 10): Minor allele count used to filter variants for PCA and entering REGENIE&#39;s variants.</li><li><code>regenie_cv_folds</code> (default: 5): Number of folds for Regenie step 1. Can also be <code>loocv</code>.</li><li><code>regenie_bsize</code> (default 1000): Regenie block size.</li></ul><p>For Regenie&#39;s options see the <a href="https://rgcgithub.github.io/regenie/options/">online documentation</a>.</p></article><nav class="docs-footer"><a class="docs-footer-prevpage" href="../tutorials/ukb_run/">« WDL-GWAS on UKBiobank RAP</a><a class="docs-footer-nextpage" href="../troubleshooting/">Troobleshooting »</a><div class="flexbox-break"></div><p class="footer-message">Powered by <a href="https://github.com/JuliaDocs/Documenter.jl">Documenter.jl</a> and the <a href="https://julialang.org/">Julia Programming Language</a>.</p></nav></div><div class="modal" id="documenter-settings"><div class="modal-background"></div><div class="modal-card"><header class="modal-card-head"><p class="modal-card-title">Settings</p><button class="delete"></button></header><section class="modal-card-body"><p><label class="label">Theme</label><div class="select"><select id="documenter-themepicker"><option value="auto">Automatic (OS)</option><option value="documenter-light">documenter-light</option><option value="documenter-dark">documenter-dark</option><option value="catppuccin-latte">catppuccin-latte</option><option value="catppuccin-frappe">catppuccin-frappe</option><option value="catppuccin-macchiato">catppuccin-macchiato</option><option value="catppuccin-mocha">catppuccin-mocha</option></select></div></p><hr/><p>This document was generated with <a href="https://github.com/JuliaDocs/Documenter.jl">Documenter.jl</a> version 1.14.1 on <span class="colophon-date" title="Friday 3 October 2025 11:14">Friday 3 October 2025</span>. Using Julia version 1.11.7.</p></section><footer class="modal-card-foot"></footer></div></div></div></body></html>
