<!DOCTYPE html>
<html lang="en"><head><meta charset="UTF-8"/><meta name="viewport" content="width=device-width, initial-scale=1.0"/><title>Workflow&#39;s Outputs · WDL-GWAS</title><meta name="title" content="Workflow&#39;s Outputs · WDL-GWAS"/><meta property="og:title" content="Workflow&#39;s Outputs · WDL-GWAS"/><meta property="twitter:title" content="Workflow&#39;s Outputs · WDL-GWAS"/><meta name="description" content="Documentation for WDL-GWAS."/><meta property="og:description" content="Documentation for WDL-GWAS."/><meta property="twitter:description" content="Documentation for WDL-GWAS."/><meta property="og:url" content="https://olivierlabayle.github.io/WDL-GWAS/stable/outputs/"/><meta property="twitter:url" content="https://olivierlabayle.github.io/WDL-GWAS/stable/outputs/"/><link rel="canonical" href="https://olivierlabayle.github.io/WDL-GWAS/stable/outputs/"/><script data-outdated-warner src="../assets/warner.js"></script><link href="https://cdnjs.cloudflare.com/ajax/libs/lato-font/3.0.0/css/lato-font.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/juliamono/0.050/juliamono.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/6.4.2/css/fontawesome.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/6.4.2/css/solid.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/6.4.2/css/brands.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/KaTeX/0.16.8/katex.min.css" rel="stylesheet" type="text/css"/><script>documenterBaseURL=".."</script><script src="https://cdnjs.cloudflare.com/ajax/libs/require.js/2.3.6/require.min.js" data-main="../assets/documenter.js"></script><script src="../search_index.js"></script><script src="../siteinfo.js"></script><script src="../../versions.js"></script><link class="docs-theme-link" rel="stylesheet" type="text/css" href="../assets/themes/catppuccin-mocha.css" data-theme-name="catppuccin-mocha"/><link class="docs-theme-link" rel="stylesheet" type="text/css" href="../assets/themes/catppuccin-macchiato.css" data-theme-name="catppuccin-macchiato"/><link class="docs-theme-link" rel="stylesheet" type="text/css" href="../assets/themes/catppuccin-frappe.css" data-theme-name="catppuccin-frappe"/><link class="docs-theme-link" rel="stylesheet" type="text/css" href="../assets/themes/catppuccin-latte.css" data-theme-name="catppuccin-latte"/><link class="docs-theme-link" rel="stylesheet" type="text/css" href="../assets/themes/documenter-dark.css" data-theme-name="documenter-dark" data-theme-primary-dark/><link class="docs-theme-link" rel="stylesheet" type="text/css" href="../assets/themes/documenter-light.css" data-theme-name="documenter-light" data-theme-primary/><script src="../assets/themeswap.js"></script></head><body><div id="documenter"><nav class="docs-sidebar"><a class="docs-logo" href="../"><img src="../assets/logo.png" alt="WDL-GWAS logo"/></a><div class="docs-package-name"><span class="docs-autofit"><a href="../">WDL-GWAS</a></span></div><button class="docs-search-query input is-rounded is-small is-clickable my-2 mx-auto py-1 px-2" id="documenter-search-query">Search docs (Ctrl + /)</button><ul class="docs-menu"><li><a class="tocitem" href="../">Home</a></li><li><span class="tocitem">Tutorials</span><ul><li><a class="tocitem" href="../tutorials/local_run/">WDL-GWAS Locally</a></li><li><a class="tocitem" href="../tutorials/ukb_run/">WDL-GWAS on UKBiobank RAP</a></li></ul></li><li><a class="tocitem" href="../inputs/">Workflow&#39;s Inputs</a></li><li class="is-active"><a class="tocitem" href>Workflow&#39;s Outputs</a><ul class="internal"><li><a class="tocitem" href="#GWAS-Summary-Statistics"><span>GWAS Summary Statistics</span></a></li><li><a class="tocitem" href="#Finemapping-Summary-Statistics"><span>Finemapping Summary Statistics</span></a></li><li><a class="tocitem" href="#Plots"><span>Plots</span></a></li><li><a class="tocitem" href="#Meta-Analysis"><span>Meta Analysis</span></a></li><li><a class="tocitem" href="#Results-Location"><span>Results Location</span></a></li></ul></li><li><a class="tocitem" href="../troubleshooting/">Troobleshooting</a></li><li><a class="tocitem" href="../contributing/">Contributing and Reporting</a></li></ul><div class="docs-version-selector field has-addons"><div class="control"><span class="docs-label button is-static is-size-7">Version</span></div><div class="docs-selector control is-expanded"><div class="select is-fullwidth is-size-7"><select id="documenter-version-selector"></select></div></div></div></nav><div class="docs-main"><header class="docs-navbar"><a class="docs-sidebar-button docs-navbar-link fa-solid fa-bars is-hidden-desktop" id="documenter-sidebar-button" href="#"></a><nav class="breadcrumb"><ul class="is-hidden-mobile"><li class="is-active"><a href>Workflow&#39;s Outputs</a></li></ul><ul class="is-hidden-tablet"><li class="is-active"><a href>Workflow&#39;s Outputs</a></li></ul></nav><div class="docs-right"><a class="docs-navbar-link" href="https://github.com/olivierlabayle/WDL-GWAS" title="View the repository on GitHub"><span class="docs-icon fa-brands"></span><span class="docs-label is-hidden-touch">GitHub</span></a><a class="docs-navbar-link" href="https://github.com/olivierlabayle/WDL-GWAS/blob/main/docs/src/outputs.md" title="Edit source on GitHub"><span class="docs-icon fa-solid"></span></a><a class="docs-settings-button docs-navbar-link fa-solid fa-gear" id="documenter-settings-button" href="#" title="Settings"></a><a class="docs-article-toggle-button fa-solid fa-chevron-up" id="documenter-article-toggle-button" href="javascript:;" title="Collapse all docstrings"></a></div></header><article class="content" id="documenter-page"><h1 id="Workflow&#39;s-Outputs"><a class="docs-heading-anchor" href="#Workflow&#39;s-Outputs">Workflow&#39;s Outputs</a><a id="Workflow&#39;s-Outputs-1"></a><a class="docs-heading-anchor-permalink" href="#Workflow&#39;s-Outputs" title="Permalink"></a></h1><p>The workflow generates a (potentially large) number of results files that can be a bit difficult to navigate at first.</p><p>In what follows, <code>GROUP_NAME</code> represents the value of a group if the <code>groupby</code> option was selected. If no groups are provided, the workflow sets it to <code>all</code>. Similarly <code>PHENOTYPE</code> is one of the provided <code>phenotypes</code> input.</p><h2 id="GWAS-Summary-Statistics"><a class="docs-heading-anchor" href="#GWAS-Summary-Statistics">GWAS Summary Statistics</a><a id="GWAS-Summary-Statistics-1"></a><a class="docs-heading-anchor-permalink" href="#GWAS-Summary-Statistics" title="Permalink"></a></h2><p>The GWAS summary statistics are stored in files with names <code>GROUP_NAME.PHENOTYPE.gwas.tsv</code>. One for each group and phenotype.</p><p>Each output file stores variant-level information and has the following columns:</p><ul><li>CHROM: The chromosome ID.</li><li>POS: The variant position.</li><li>ID: The variant ID.</li><li>ALLELE_0: The baseline allele. This should be the REF allele provided in the input PGEN files.</li><li>ALLELE_1: The effect allele, this should be the ALT allele provided in the input PGEN files.</li><li>ALLELE<em>1</em>FREQ: The ALLELE_1 frequency.</li><li>BETA: The effect size.</li><li>SE: The standard error.</li><li>LOG10P: The negative log10(p-value).</li><li>N: The number of samples used for this variant.</li></ul><p>Furthermore:</p><ul><li><strong>SAIGE</strong> outputs the extra following: ALLELE<em>1</em>COUNT, MISSING<em>RATE, T</em>STAT, VAR.</li><li><strong>REGENIE</strong> outputs theextra following: TEST, CHISQ, EXTRA.</li></ul><h2 id="Finemapping-Summary-Statistics"><a class="docs-heading-anchor" href="#Finemapping-Summary-Statistics">Finemapping Summary Statistics</a><a id="Finemapping-Summary-Statistics-1"></a><a class="docs-heading-anchor-permalink" href="#Finemapping-Summary-Statistics" title="Permalink"></a></h2><p>Similarly to GWAS results, finemapping summary statistics are stored in files with names <code>GROUP_NAME.PHENOTYPE.finemapping.tsv</code>. One for each group and phenotype.</p><p>The columns of those files are:</p><ul><li>CHROM: The chromosome ID.</li><li>POS: The variant position.</li><li>ID: The variant ID.</li><li>REF: The reference allele from the input PGEN file</li><li>ALT: The alternate allele from the input PGEN file</li><li>LOCUS_ID: The GWAS lead variant ID identifying the locus.	</li><li>PIP: The posterior inclusion probability.</li><li>CS: The credible set ID.	</li><li>UNPHASED_R2: The R2 with the GWAS lead variant ID.</li><li>SUSIE_CONVERGED: Whether the SuSiE algorithm converged.</li></ul><h2 id="Plots"><a class="docs-heading-anchor" href="#Plots">Plots</a><a id="Plots-1"></a><a class="docs-heading-anchor-permalink" href="#Plots" title="Permalink"></a></h2><p>Finally, the workflow outputs a series of plots:</p><ul><li>The GWAS plots, formatted as <code>GROUP_NAME.PHENOTYPE.PLOT_TYPE.png</code>. One for each group, phenotype and plot type (currently QQ, Manhattan).</li><li>Optional locus plots, formatted as <code>GROUP_NAME.PHENOTYPE.LEAD_VARIANT_ID.locuszoom.png</code>. One for each group, phenotype and locus passing the fine-mapping thresholds.</li></ul><h2 id="Meta-Analysis"><a class="docs-heading-anchor" href="#Meta-Analysis">Meta Analysis</a><a id="Meta-Analysis-1"></a><a class="docs-heading-anchor-permalink" href="#Meta-Analysis" title="Permalink"></a></h2><p>If the <code>groupby</code> and <code>meta_analysis</code> options are specified, the workflow will also produce meta-analysis outputs. These are similar to the above specified outputs with <code>GROUP_NAME=META_ANALYSIS</code>.</p><p>The GWAS output fill then contain the following additional columns:</p><ul><li>DIRECTION: Summary of effect direction for each study, with one &#39;+&#39; or &#39;-&#39; per study</li><li>HET_ISQ: I^2 statistic which measures heterogeneity on scale of 0-100%</li><li>HET_CHISQ: Chi-squared statistic in simple test of heterogeneity</li><li>HET_DF: Degrees of freedom for heterogeneity statistic</li><li>LOG10P_HET: log10 of p-value for heterogeneity statistic</li></ul><h2 id="Results-Location"><a class="docs-heading-anchor" href="#Results-Location">Results Location</a><a id="Results-Location-1"></a><a class="docs-heading-anchor-permalink" href="#Results-Location" title="Permalink"></a></h2><p>All results are found in the output folder whose lcoation and organisation depend on the execution engine.</p><h3 id="Local-Execution-with-Cromwell"><a class="docs-heading-anchor" href="#Local-Execution-with-Cromwell">Local Execution with Cromwell</a><a id="Local-Execution-with-Cromwell-1"></a><a class="docs-heading-anchor-permalink" href="#Local-Execution-with-Cromwell" title="Permalink"></a></h3><p>The outputs will be stored in the output folder defined by the <code>final_workflow_outputs_dir</code> variable specified in the cromwell&#39;s <a href="https://cromwell.readthedocs.io/en/latest/wf_options/Overview/">option file</a>. The output folder is typically organised in subfolders corresponding to the workflow&#39;s tasks.</p><p>For example, for a grouped analysis with meta-analysis you should have the following subfolders:</p><ul><li><code>call-gwas_group_plots</code>: Containing plots for each analysed group.</li><li><code>call-gwas_meta_plots</code>: Containing the plots for the meta-analysis.</li><li><code>call-merge_gwas_group_chr_results</code>: Containing GWAS results for each group.</li><li><code>call-meta_analyse</code>: Containing meta-analysed GWAS results.</li><li><code>call-merge_fp_group_chr_results</code>: Containing finemapping results for each group.</li><li><code>call-merge_fp_meta_chr_results</code>: Containing finemapping results from the meta-analysis stage using summary statistics.</li></ul><h3 id="UK-Biobank-RAP"><a class="docs-heading-anchor" href="#UK-Biobank-RAP">UK Biobank RAP</a><a id="UK-Biobank-RAP-1"></a><a class="docs-heading-anchor-permalink" href="#UK-Biobank-RAP" title="Permalink"></a></h3><p>In that case, the output folder is defined by the <code>destination</code> argument provided to <code>dx run</code>. Results will be found as a flat list in the output folder and all intermediate results are stored in the <code>intermediate</code> subfolder.</p></article><nav class="docs-footer"><a class="docs-footer-prevpage" href="../inputs/">« Workflow&#39;s Inputs</a><a class="docs-footer-nextpage" href="../troubleshooting/">Troobleshooting »</a><div class="flexbox-break"></div><p class="footer-message">Powered by <a href="https://github.com/JuliaDocs/Documenter.jl">Documenter.jl</a> and the <a href="https://julialang.org/">Julia Programming Language</a>.</p></nav></div><div class="modal" id="documenter-settings"><div class="modal-background"></div><div class="modal-card"><header class="modal-card-head"><p class="modal-card-title">Settings</p><button class="delete"></button></header><section class="modal-card-body"><p><label class="label">Theme</label><div class="select"><select id="documenter-themepicker"><option value="auto">Automatic (OS)</option><option value="documenter-light">documenter-light</option><option value="documenter-dark">documenter-dark</option><option value="catppuccin-latte">catppuccin-latte</option><option value="catppuccin-frappe">catppuccin-frappe</option><option value="catppuccin-macchiato">catppuccin-macchiato</option><option value="catppuccin-mocha">catppuccin-mocha</option></select></div></p><hr/><p>This document was generated with <a href="https://github.com/JuliaDocs/Documenter.jl">Documenter.jl</a> version 1.14.1 on <span class="colophon-date" title="Friday 21 November 2025 18:42">Friday 21 November 2025</span>. Using Julia version 1.11.7.</p></section><footer class="modal-card-foot"></footer></div></div></div></body></html>
